### PR TITLE
[WIP] Speculative decoding using a draft model

### PR DESCRIPTION
**Pull request by cadedaniel**
### Speculative decoding
This PR adds speculative decoding to vLLM using the draft model approach. This was first explored in these papers:
* [Accelerating Large Language Model Decoding with Speculative Sampling](https://arxiv.org/abs/2302.01318)
* [Fast Inference from Transformers via Speculative Decoding](https://arxiv.org/abs/2211.17192)

As a simplified overview, this type of speculative decoding runs a smaller draft model to guess what the larger target model will emit. The target model then verifies the guesses, and emits them if they pass verification. This yields a latency reduction as the verification of many tokens can happen in a single forward pass of the target model.

Running this at Anyscale, I see a 30-50% latency reduction depending on the draft model, target model and dataset.

### Usage
The usage looks like this:
```
llm = LLM(
    model="meta-llama/Llama-2-13b-chat-hf",
    speculative_model="JackFram/llama-68m", # The draft model. Must have same vocabulary as target model.
    tensor_parallel_size=4,
    speculative_model_uses_tp_1=True, # Whether the draft model should use TP=1 or same TP as target model.
    num_speculative_tokens=3, # The number of speculative tokens to score.
)
```

### Feature list
* Any vLLM model can be used as the draft model as long as it has the same vocabulary as the target model.
* Rejection sampling to guarantee target model distribution.
* The draft model can use tensor-parallel degree of 1, or the same tensor-parallel degree as the target model. This reduces draft model latency and contributes to the overall speedup.
* The vLLM scheduler has been improved to support >1 token per step.
* Tests for rejection sampling, various workers, scheduler, and e2e speculative decoding.

### Future work (not implemented)
* Top-k speculative decoding ("Tree attention") to increase draft hitrates
* Optimized PagedAttention (multi-query attention) for scoring
* Beam-sampling + speculative decoding

This PR is marked as draft as there is nontrivial work required to get this into a mergable state.

### Guide for reviewers
The following are key files for understanding this speculative decoding implementation:
* [Rejection sampler](https://github.com/cadedaniel/vllm-public/blob/301f00d62cd37ce04b093a8c4614f5a11793d599/vllm/model_executor/layers/rejection_sampler.py). Use [rejection sampling](https://en.wikipedia.org/wiki/Rejection_sampling) to approximate the target distribution using samples from the draft distribution.
* [Draft target worker](https://github.com/cadedaniel/vllm-public/blob/301f00d62cd37ce04b093a8c4614f5a11793d599/vllm/worker/draft_target_worker.py). A worker which contains both the draft and target models. This orchestrates the drafting of speculative continuations, scoring, and accepting speculative tokens based on rejection sampling.
* [Multi-step worker](https://github.com/cadedaniel/vllm-public/blob/301f00d62cd37ce04b093a8c4614f5a11793d599/vllm/worker/multi_step_worker.py). This runs the draft model several times, without invoking the vLLM scheduler each time.
* [Single-TP worker](https://github.com/cadedaniel/vllm-public/blob/301f00d62cd37ce04b093a8c4614f5a11793d599/vllm/worker/single_tp_worker.py). This allows the DraftTargetWorker to run the draft model with TP=1, while the target model is TP=2, 4 or 8. This reduces latency of the draft model.
* [Key scheduler change](https://github.com/cadedaniel/vllm-public/blob/301f00d62cd37ce04b093a8c4614f5a11793d599/vllm/core/scheduler.py#L221-L227). The scheduler and block manager now have the notion of "preallocated slots". This allows them to schedule KV block space sufficient for the worker to run several steps before the next scheduler invocation.
* [Key sequence changes](https://github.com/cadedaniel/vllm-public/blob/301f00d62cd37ce04b093a8c4614f5a11793d599/vllm/sequence.py#L130-L143). Sequences now track "processed tokens". A processed token is one that has been processed by the model such that the KV activations have been saved to cache.
* Performance optimizations. Significant modifications have been made to the sampler so that the draft model can run several steps without CPU synchronization. This plus the SingleTPWorker reduced draft model latency from 5ms to 1.5ms.

### CODE REVIEW

**Comment by Lvjinhong** [comment]
How is the current progress? May I use your method to accelerate the llama2 70b on vLLM for now?

**Comment by cadedaniel** [comment]
> I'm very fortunate to witness such great work from you. How is the current progress? May I use your method to accelerate the llama2 70b on vLLM for now?

hi @Lvjinhong. The current PR requires some work to get into a working state on the public vllm repo. I will start on this later this week but given US holidays I expect to finish early january.

**Comment by zhaoyang-star** [comment]
Very glad to see this work. I also have a wip version of speculative decoding and looking forward to using this feature.

**Comment by cadedaniel** [comment]
Created a plan to break this PR into separate pieces. Pending review from vLLM original authors, I will start on it this week. https://docs.google.com/document/d/1rE4pr3IdspRw97XbImY4fS9IWYuJJ3HGtL7AdIKGrw8/edit

**Comment by wasertech** [comment]
So from the little I understood I think there is a way to avoid having a draft model for your vocabulary by using directly the model's ngrams? It should probably be the subject of a separate PR but I thinks this is the way forward for everyone to easily get the benefits of speculative decoding.

What do you think @cadedaniel, do you think its as straigh forward as @simon-mo puts it?
> I would imagine after this PR, ngram support should be very straightforward.

**Comment by cadedaniel** [comment]
Yep. This PR builds the framework for scoring and verifying draft tokens, independent of whether they come from a draft model or something like Medusa, or lookup ngrams like prompt lookup or RAG lookup.

**Comment by wasertech** [comment]
I think I understand a bit better know thank you, @cadedaniel , again for this incredible work!

**Comment by UranusSeven** [comment]
@cadedaniel Hi! Thanks for your great work!

As I understand it, in each iteration, we are going to run the draft model several times and the target model once. Will this increase the first token latency greatly?

**Comment by cadedaniel** [comment]
Yes, the time to first token is a few milliseconds higher for a draft model of size 68m. It can be optimized in future versions, e.g. with Medusa/EAGLE where draft tokens are generated without independent kv cache.

**Comment by UranusSeven** [comment]
@cadedaniel Thanks for your reply! I also want to confirm my understanding regarding the decoding step's impact on first token latency:

- Generating `k` draft tokens per iteration introduces a delay of `k * d` milliseconds, where `d` is the time to run the draft model once.
- Additionally, the target model's execution time of `t` milliseconds further contributes to the overall latency.
- This means the total decoding time is approximately `k * d + t` milliseconds. The cost of sampling is ignored.
- Consequently, newly arriving prefill requests might experience a wait of up to `k * d + t` milliseconds before execution.

I'm wondering if it helps to break the decoding step into 2 sub-steps, drafting and verifying. In this way, newly arriving prefill requests wait up to `k * d` milliseconds. And, if the newly arriving prompts are short, they can be batched with the verifying step.

What do you think?

**Comment by cadedaniel** [comment]
@UranusSeven could you ask your question in a discussion post? happy to answer there

**Comment by SinanAkkoyun** [comment]
Hi! Do you have some speed tps benchmarks?

**Comment by qizzzh** [comment]
Out of curiosity, does the proposal support separate deployment for the draft and target model? Asking because in production these two likely have different QPS and computing resource requirements.

File: vllm/config.py [hunk 1]
```diff
@@ -49,31 +51,32 @@ class ModelConfig:
...
+                 seed: int,
+                 revision: Optional[str],
+                 tokenizer_revision: Optional[str] = None,
+                 max_model_len: Optional[int] = None,
+                 quantization: Optional[str] = None,
+                 enable_cuda_graph: bool = False,
+                 cuda_graph_max_context_len: int = 5000,
+                 cuda_graph_cache_size: int = 10,
+                 flash_style: bool = False,
+                 max_chunked_prefill_len: int = -1,
```
Comment (xunfeng1980) [inline vllm/config.py:79]: ) ?

**Review by xunfeng1980 — COMMENTED** [review]
```
vllm-public/vllm/config.py", line 62
    def __init__(self,
                ^
SyntaxError: '(' was never closed

````

File: vllm/config.py (hunk 1 above)
Comment (xunfeng1980) [reply vllm/config.py:79]:
```
vllm-public/vllm/engine/async_llm_engine.py", line 7, in <module>
    from vllm.anyscale.lora.utils import LoRARequest
ModuleNotFoundError: No module named 'vllm.anyscale'
```

**Comment by xunfeng1980** [comment]
> This PR is about drinking fake wine?

**Comment by cadedaniel** [comment]
Hi everyone, want to provide a quick update here:
* Last few weeks I've prioritized optimizing Mixtral latency https://github.com/vllm-project/vllm/pull/2913
* Now I will focus on getting this merged full-time. I am aiming to finish merges by February with @LiuXiaoxuanPKU's help reviewing.
* After the correctness tests are merged, I will accept any optimizations that pass correctness tests. I will list out some major optimizations that people can take on (and already some tech discussions happening on MQA cc @ymwangg @robertgshaw2-neuralmagic ).

**Comment by binarycrayon** [comment]
Thank you for your update and your efforts!

**Comment by ymwangg** [comment]
Thanks for the update! Looking forward to seeing how to incorporate our work once your PRs are out.

**Comment by hchoi-moveworks** [comment]
Thanks @cadedaniel for truly inspiring work! 🙏 

Would Speculative decoding work with vLLM's continuous batching as well? Would that be the step `Integrate speculative decoding with LLMEngine` in our proposed design doc? 
https://docs.google.com/document/d/1rE4pr3IdspRw97XbImY4fS9IWYuJJ3HGtL7AdIKGrw8/edit

**Comment by cadedaniel** [comment]
> Thanks @cadedaniel for truly inspiring work! 🙏 
> 
> 
> 
> Would Speculative decoding work with vLLM's continuous batching as well? Would that be the step `Integrate speculative decoding with LLMEngine` in our proposed design doc? 
> 
> https://docs.google.com/document/d/1rE4pr3IdspRw97XbImY4fS9IWYuJJ3HGtL7AdIKGrw8/edit

Yep! Once the e2e correctness tests pass you can use it with continuous batching.

**Comment by arunpatala** [comment]
Thanks for the great work you are doing. I wonder what is the timeline looking like. thanks

File: vllm/config.py (hunk 1 above)
Comment (vitali87) [reply vllm/config.py:79]: Having the same issue. Any update on this?

**Comment by paolovic** [comment]
Hi @cadedaniel ,
thanks for your great contribution!
Since I'm also very keen on this feature, can I support you somehow?

**Comment by cadedaniel** [comment]
Thanks! It would be helpful if you could test it out and add features or optimizations. Once https://github.com/vllm-project/vllm/pull/3951 is merged it will be correct but not yet fast; I'll post more about optimizing it this week.

**Comment by cocoza4** [comment]
will `AssertionError: Speculative decoding not yet supported for RayGPU backend.` error go away if this PR is merged? Ref: https://github.com/vllm-project/vllm/issues/4358

**Comment by HimanshuJanbandhu** [comment]
Hi guys, 
First of all great to see your work on speculative decoding.
However, there is a new paper, Draft & Verify: Lossless Large Language Model Acceleration via
Self-Speculative Decoding (https://arxiv.org/abs/2309.08168), which talks about how we can create a draft model, by skipping layers from the original LLM.
There's another paper recently by Meta, that talks about it, LayerSkip: Enabling Early Exit Inference and
Self-Speculative Decoding - https://arxiv.org/pdf/2404.16710.
This method leads further optimization of memory as it is using the same model again. Would love to see this implemented in vLLM, as I am keen on this feature, can I support you somehow?

**Comment by w32zhong** [comment]
Hi @HimanshuJanbandhu, I would like also mention my recent work S3D (https://arxiv.org/abs/2405.20314). It is very similar to your mentioned Self-Speculative Decoding work which is simple to implement and easy to be integrated to existing stacks. But we have achieved better efficiency in general (compared to Self-Spec), our method combines layer-skipping with multiple next-token generation/unmasking. Although ours requires a bit training, but it should be straightforward just like training a Transformer decoder like BERT.

**Comment by cadedaniel** [comment]
We welcome a self-speculative implementation!

**Comment by paolovic** [comment]
> > Thanks! It would be helpful if you could test it out and add features or optimizations. Once #3951 is merged it will be correct but not yet fast; I'll post more about optimizing it this week.
> 
> Great, I'll start tomorrow (CET)

by the way: sorry for volunteering and never coming back to you. Unfortunately, I am and will be reaaally busy until beginning of August ✌🏻

**Comment by Lisiqi-a** [comment]
Hello, Teacher, it is a great honor to witness your magnificent work. I have been studying you SpecDecode  work recently, and I have a question that I hope you can guide me on. Why does SpecDecodeWorker inherit from LoraNotSupportedWorkerBase? Why doesn't the current SpecDecode support Lora? @cadedaniel 
screenshot-20240805-113310

**Comment by cadedaniel** [comment]
Hi @skylee-01 . See https://github.com/vllm-project/vllm/issues/6912 for the work required to add LoRA + spec decode.

**Comment by cadedaniel** [comment]
Thanks for the interest everyone! With https://github.com/vllm-project/vllm/issues/4630#issuecomment-2269618262, all the work in this PR has been merged.

**Pull request closed without merging**